test/persistence-hooks.suite.js

```diff
@@ -2946,6 +2946,372 @@ module.exports = function(dataSource, should, connectorCapabilities) {
+            { id: 'ignored', name: 'new name' },
+            function(err, instance) {
+              if (err) return done(err);
+              hookMonitor.names.should.eql(['access', 'before save']);
+              done();
+            });
+      });
+
+      it('applies updates from `access` hook when found', function(done) {
+        TestModel.observe('access', function(ctx, next) {
+          ctx.query = { where: { id: { neq: existingInstance.id }}};
+          next();
+        });
+
+        TestModel.upsertWithWhere({ id: existingInstance.id },
+            { name: 'new name' },
+            function(err, instance) {
+              if (err) return done(err);
+              findTestModels({ fields: ['id', 'name'] }, function(err, list) {
+                if (err) return done(err);
+                (list || []).map(toObject).should.eql([
+                  { id: existingInstance.id, name: existingInstance.name, extra: undefined },
+                  { id: instance.id, name: 'new name', extra: undefined },
+                ]);
+                done();
+              });
+            });
+      });
+
+      it('applies updates from `access` hook when not found', function(done) {
+        TestModel.observe('access', function(ctx, next) {
+          ctx.query = { where: { id: 'not-found' }};
+          next();
+        });
+
+        TestModel.upsertWithWhere({ id: existingInstance.id },
+            { name: 'new name' },
+            function(err, instance) {
+              if (err) return done(err);
+              findTestModels({ fields: ['id', 'name'] }, function(err, list) {
+                if (err) return done(err);
+                (list || []).map(toObject).should.eql([
+                  { id: existingInstance.id, name: existingInstance.name, extra: undefined },
+                  { id: list[1].id, name: 'second', extra: undefined },
+                  { id: instance.id, name: 'new name', extra: undefined },
+                ]);
+                done();
+              });
+            });
+      });
+
+      it('triggers `before save` hook on update', function(done) {
+        TestModel.observe('before save', ctxRecorder.recordAndNext());
+
+        TestModel.upsertWithWhere({ id: existingInstance.id },
+            { id: existingInstance.id, name: 'updated name' },
+            function(err, instance) {
+              if (err) return done(err);
+              if (dataSource.connector.upsertWithWhere) {
+                ctxRecorder.records.should.eql(aCtxForModel(TestModel, {
+                  where: { id: existingInstance.id },
+                  data: { id: existingInstance.id, name: 'updated name' },
+                }));
+              } else {
+                ctxRecorder.records.should.eql(aCtxForModel(TestModel, {
+                  where: { id: existingInstance.id },
+                  data: { id: existingInstance.id, name: 'updated name' },
+                  currentInstance: existingInstance,
+                }));
```

Comment:
The expected objects if the implementation of `dataSource.connector.upsertWithWhere` exists or not would be simillar; the only difference is that for the latter `data` object has `currentInstance` as well? In this case I think it is cleaner if we have some code resuability; one object for both and just append `currentInstance` to `data` object if `dataSource.connector.upsertWithWhere` is false